core/src/main/java/io/questdb/cairo/sql/async/PageFrameSequence.java

```diff
@@ -0,0 +1,440 @@
+    private static final int OWNER_NONE = 0;
+    private static final int OWNER_WORK_STEALING = 1;
+    private static final int OWNER_ASYNC = 2;
+    private final static Log LOG = LogFactory.getLog(PageFrameSequence.class);
+    private static final AtomicLong ID_SEQ = new AtomicLong();
+    public final SOUnboundedCountDownLatch doneLatch = new SOUnboundedCountDownLatch();
+    private final AtomicBoolean valid = new AtomicBoolean(true);
+    private final AtomicInteger reduceCounter = new AtomicInteger(0);
+    private final LongList frameRowCounts = new LongList();
+    private final PageFrameReducer reducer;
+    private final PageAddressCache pageAddressCache;
+    private final MessageBus messageBus;
+    private final AtomicInteger owner = new AtomicInteger(OWNER_NONE);
+    private final MicrosecondClock microsecondClock;
+    private long id;
+    private int shard;
+    private int frameCount;
+    private Sequence collectSubSeq;
+    private SymbolTableSource symbolTableSource;
+    private T atom;
+    private PageAddressCacheRecord[] records;
+    // we need this to restart execution for `toTop`
+    private MPSequence dispatchPubSeq;
+    private RingQueue<PageFrameDispatchTask> pageFrameDispatchQueue;
+    private final AtomicInteger dispatchStartIndex = new AtomicInteger();
+    private SqlExecutionCircuitBreaker[] circuitBreakers;
+    private long startTimeUs;
+    private long circuitBreakerFd;
+
+    public PageFrameSequence(
+            CairoConfiguration configuration,
+            MessageBus messageBus,
+            PageFrameReducer reducer
+    ) {
+        this.pageAddressCache = new PageAddressCache(configuration);
+        this.messageBus = messageBus;
+        this.reducer = reducer;
+        this.microsecondClock = configuration.getMicrosecondClock();
+    }
+
+    /**
+     * Waits for frame sequence completion, fetches remaining pieces of the
+     * frame sequence from the queues. This method is not thread safe.
+     */
+    public void await() {
+        LOG.debug()
+                .$("awaiting completion [shard=").$(shard)
+                .$(", id=").$(id)
+                .$(", frameCount=").$(frameCount)
+                .I$();
+
+        // Collect only dispatched frames. To do that we need to stall any further dispatch.
+        while (true) {
+            final int dispatched = dispatchStartIndex.get();
+            if (dispatchStartIndex.compareAndSet(dispatched, frameCount)) {
+                frameCount = dispatched;
+                break;
+            }
+        }
+
+        final RingQueue<PageFrameReduceTask> queue = getPageFrameReduceQueue();
+        final MCSequence pageFrameReduceSubSeq = messageBus.getPageFrameReduceSubSeq(shard);
+        while (doneLatch.getCount() == 0) {
+            final int workerId = getWorkerId();
+            final PageAddressCacheRecord rec = records[workerId];
+            final SqlExecutionCircuitBreaker circuitBreaker = circuitBreakers[getWorkerId()];
```

Comment:
nit: should be
```suggestion
            final SqlExecutionCircuitBreaker circuitBreaker = circuitBreakers[workerId];
```

Reply:
done in https://github.com/questdb/questdb/pull/1732/commits/40648e75d7590eb17e3526b9b0cc362e1754a925